pkg/api/bzz.go

```diff
@@ -252,81 +398,83 @@ FETCH:
 	}
 
 	// serve requested path
-	s.serveManifestEntry(w, r, address, me.Reference(), !feedDereferenced)
+	s.serveManifestEntry(w, r, address, me, !feedDereferenced)
 }
 
-func (s *server) serveManifestEntry(w http.ResponseWriter, r *http.Request, address, manifestEntryAddress swarm.Address, etag bool) {
-	var (
-		logger = tracing.NewLoggerWithTraceID(r.Context(), s.logger)
-		ctx    = r.Context()
-		buf    = bytes.NewBuffer(nil)
-	)
-
-	// read file entry
-	j, _, err := joiner.New(ctx, s.storer, manifestEntryAddress)
-	if err != nil {
-		logger.Debugf("bzz download: joiner read file entry %s: %v", address, err)
-		logger.Errorf("bzz download: joiner read file entry %s", address)
-		jsonhttp.NotFound(w, nil)
-		return
+func (s *server) serveManifestEntry(
+	w http.ResponseWriter,
+	r *http.Request,
+	address swarm.Address,
+	manifestEntry manifest.Entry,
+	etag bool,
+) {
+
+	additionalHeaders := http.Header{}
+	mtdt := manifestEntry.Metadata()
+	if fname, ok := mtdt[manifestEntryMetadataFilenameKey]; ok {
+		additionalHeaders["Content-Disposition"] =
+			[]string{fmt.Sprintf("inline; filename=\"%s\"", fname)}
 	}
-
-	_, err = file.JoinReadAll(ctx, j, buf)
-	if err != nil {
-		logger.Debugf("bzz download: read file entry %s: %v", address, err)
-		logger.Errorf("bzz download: read file entry %s", address)
-		jsonhttp.NotFound(w, nil)
-		return
+	if mimeType, ok := mtdt[manifestEntryMetadataContentTypeKey]; ok {
+		additionalHeaders["Content-Type"] = []string{mimeType}
 	}
-	fe := &entry.Entry{}
-	err = fe.UnmarshalBinary(buf.Bytes())
-	if err != nil {
-		logger.Debugf("bzz download: unmarshal file entry %s: %v", address, err)
-		logger.Errorf("bzz download: unmarshal file entry %s", address)
-		jsonhttp.NotFound(w, nil)
-		return
+
+	s.downloadHandler(w, r, manifestEntry.Reference(), additionalHeaders, etag)
+}
+
+// downloadHandler contains common logic for dowloading Swarm file from API
+func (s *server) downloadHandler(w http.ResponseWriter, r *http.Request, reference swarm.Address, additionalHeaders http.Header, etag bool) {
```

Comment:
just a minor comment, I think that `downloadHandler` is not used anywhere else except `serveManifestEntry`, so I'm not sure we need it as an extra method. Up to you

Reply:
Its actually required in bytes file also.